dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+  `ILogger.BeginScope` or similar.
+
+### Async Completion
+
+Simply `throw new Temporal.Activity.CompleteAsyncException` from the activity.
+
+## Client
+
+Inside the `Temporal.Client` namespace, we have the following types:
+
+* `TemporalClient` - namespace-specific, stateless high-level client to Temporal
+  * See later for most of its API design
+  * Accepts `ITemporalConnection`
+  * Implements `ITemporalClient` for DI use (which itself implements `Temporal.Worker.IWorkerClient` for worker use)
+* `TemporalConnection` - namespace-agnostic, stateful low-level client to Temporal
+  * Has `WorkflowService`, `OperatorService`, etc for making raw gRPC calls
+  * Supports lazy connectivity
+  * Implements `ITemporalConnection` for DI use
+* `WorkflowHandle` - high-level workflow-specific calls
+* Many other supporting types
+
+Example of starting the workflow defined earlier in document, signalling it, then waiting on its results:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Start workflow
+var handle = await client.StartWorkflowAsync(
+    IGreetingWorkflow.Ref.RunAsync, "Temporal", new(ID: "my-workflow-id", TaskQueue: "my-task-queue"));
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
+
+// Wait for result
+Console.WriteLine(await handle.GetResultAsync());
+```
+
+Example of doing the same via already started workflow:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Get workflow handle
+var handle = await client.GetWorkflowHandle<string>("my-workflow-id");
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
+
+// Wait for result
+Console.WriteLine(await handle.GetResultAsync());
+```
+
+Note, this could have been just `client.GetWorkflowHandle("my-workflow-id")` + `handle.GetResultAsync<string>()` if the
+user would rather not provide the type until fetching the result. See the definitions below.
+
+### TemporalClient
+
+Looks like:
+
+```csharp
+namespace Temporal.Client;
+
+public class TemporalClient : ITemporalClient
+{
+    public static async Task<TemporalClient> ConnectAsync(TemporalClientConnectOptions options) { /*...*/ }
+
+    public TemporalClient(ITemporalConnection connection, TemporalClientOptions options) { /*...*/ }
+
+    public string Namespace { get; }
+    public DataConverter DataConverter { get; }
+    public IBridgeClientProvider BridgeClientProvider { get; }
+    public ITemporalConnection Connection { get; }
+    public TemporalClientOptions Options { get; }
+
+    // StartWorkflowAsync
+
+    public async Task<WorkflowHandle<TResult>> StartWorkflowAsync<TResult>(
+        string workflow, object[] args, StartWorkflowOptions options) { /*...*/ }
```

Comment:
With the proxy lambda approach, would we be able to support type safe multi argument methods?

Reply:
Yes, but we have avoided those in Python too where we could have done a similar thing. In addition to requiring the non-interface methods be virtual, it also introduces some difficulties (e.g. I can't stop you from making other calls in the lambda either).

Reply:
Added discussion point on lambda vs method reference